torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py

```diff
@@ -0,0 +1,195 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import binascii
+import codecs
+import logging
+import os
+from datetime import timedelta
+from typing import Any, Optional, Tuple, cast
+
+from torch.distributed import Store, TCPStore
+
+from .api import RendezvousConnectionError, RendezvousParameters, RendezvousStateError
+from .default_rendezvous import RendezvousBackend, Token
+from .utils import _matches_machine_hostname, _parse_rendezvous_endpoint
+
+log = logging.getLogger(__name__)
+
+
+class C10dRendezvousBackend(RendezvousBackend):
+    """Represents a C10d-backed rendezvous backend.
+
+    Args:
+        store:
+            The `Store` instance to use to communicate with the C10d store.
+        run_id:
+            The run id of the rendezvous.
+    """
+
+    # See the explanation in the __init__ method.
+    _NULL_SENTINEL = "Y2FuaW1hZGFt"
+
+    _store: Store
+    _key: str
+
+    def __init__(self, store: Store, run_id: str) -> None:
+        if not run_id:
+            raise ValueError("The run id must be a non-empty string.")
+
+        self._store = store
+
+        self._key = "torch.rendezvous." + run_id
+
+        # The read operation of a store blocks the caller until the specified
+        # key becomes available. This behavior makes it tricky to use a store
+        # as a regular key-value dictionary.
+        #
+        # As a workaround we initially set a sentinel value as the rendezvous
+        # state. Whenever this value gets returned we treat it as a None.
+        self._call_store("compare_set", self._key, "", self._NULL_SENTINEL)
+
+    @property
+    def name(self) -> str:
+        """See base class."""
+        return "c10d-experimental"
+
+    @property
+    def store(self) -> Store:
+        """Gets the `Store` instance used to communicate with the C10d store."""
+        return self._store
+
+    @property
+    def key(self) -> str:
+        """Gets the key under which the rendezvous state is stored."""
+        return self._key
+
+    def get_state(self) -> Optional[Tuple[bytes, Token]]:
+        """See base class."""
+        base64_state: bytes = self._call_store("get", self._key)
+
+        return self._decode_state(base64_state)
+
+    def set_state(
+        self, state: bytes, token: Optional[Token] = None
+    ) -> Optional[Tuple[bytes, Token]]:
+        """See base class."""
+        base64_state_str: str = codecs.encode(state, "base64").decode()
+
+        if token:
+            # Shortcut if we know for sure that the token is not valid.
+            if not isinstance(token, bytes):
+                return self.get_state()
+
+            token = token.decode()
+        else:
+            token = self._NULL_SENTINEL
+
+        base64_state: bytes = self._call_store("compare_set", self._key, token, base64_state_str)
```

Comment:
Why do we have a call_store API like this instead of just using `self._store.compare_set(....)`?

Reply:
We call the store in several places in `C10dRendezvousBackend`. I wanted to make sure that we wrap each call with an identical try/except clause to propagate errors consistently as `RendezvousConnectionError`s (where the original error is put into `__cause__`). I could either duplicate the same try/except in all places where we call the store, or come up with a fancy decorator or context manager to "magically" wrap them, but thought that this would be the most straightforward solution.